## Notebook 4C: Seatwork 2

<b>Instructions:</b> Submit your answers by <b>LT</b>.

LT Number: 10

Names:\
&emsp;Janica Lae Batang\
&emsp;Mary Honeylie Buitre\
&emsp;Paolo Jose de Guzman\
&emsp;John Francis Gacal\
&emsp;Vincent John Rivera

## Eigenvalue Application: Markov Chains

A <b><i>Markov Chain</i></b> is a system that has $n$ possible states and passes through a series of transitions from one state to another.

The probability of a transition from state $i$ to state $j$ is given by $a_{ij}$ where

\begin{align}
    0 \leq a_{ij} \leq 1
\end{align}

\begin{align}
    \sum_{i = 1}^{n} a_{ij} = 1
\end{align}

Consider the following <b><i>stock market state model</i></b>:

<img src = "stockmarket.png" width = "300"/>

The model is typically represented as a <b><i>probability transition matrix</i></b>, given by

\begin{align}
    A = \begin{bmatrix}
        0.975 & 0.3 & 0.02\\
        0.02 & 0.5 & 0.40\\
        0.005 & 0.2 & 0.58
    \end{bmatrix}
\end{align}

Note that each column contains the probabilities of moving to state $i$ given that you are in state $j$, such that

- State $1$ corresponds to being in the Bull Market state
- State $2$ corresponds to being in the Bear Market state
- State $3$ corresponds to being in the Stagnant Market state

Some examples:
- Given that we are in a Bull Market (column 1), the probability that in the next time step we stay in a Bull Market is 97.5%.
- Given that we are in a Stagnant Market (column 3), the probability that in the next time step we enter a Bear Market is 40%.
- And so on...

Given $A$, let $x_i^{(k)}$ denote the probability that the system is in state $i$ after transition $k$:

\begin{align}
    x^{(k)} = \begin{bmatrix}
                x_1^{(k)}\\
                x_2^{(k)}\\
                x_3^{(k)}
            \end{bmatrix}
\end{align}

If the initial <b><i>probability distribution vector</i></b> is $x^{(0)}$, then the probability distribution vector after $k$ steps is

\begin{align}
    x^{(k)} = Ax^{(k-1)} = A^kx^{(0)}
\end{align}

The long-term behavior of the system is therefore determined by the value of

\begin{align}
    \lim_{k \rightarrow \infty} A^k
\end{align}

#### Q1.

Suppose that the stock market is initially in a Bull Market.

\begin{align}
    x^{(0)} = \begin{bmatrix}
                1\\
                0\\
                0\\
            \end{bmatrix}
\end{align}

What is the probability distribution vector after three steps?

In [2]:
import numpy as np
# Insert code here
A = np.array([[0.975, 0.3, 0.02],
              [0.02, 0.5, 0.4],
              [0.005, 0.2, 0.58]])

x_0 = np.array([1, 0, 0])

x_3 = np.matmul(np.linalg.matrix_power(A, 3),x_0)
x_3

array([0.94249237, 0.0395945 , 0.01791313])

#### Q2.

What is the long-term value of the probability distribution vector? (i.e. $\lim_{k \rightarrow \infty} x^{(k)}$)

In [166]:
# Insert code here
for i in range(1,100):
    x_k = np.matmul(np.linalg.matrix_power(A, i),x_0)
    if i % 5 == 0:
        print(f'i = {i} -- x_k = {x_k}')

i = 5 -- x_k = [0.92209163 0.0507379  0.02717047]
i = 10 -- x_k = [0.89741677 0.06406867 0.03851455]
i = 15 -- x_k = [0.88913757 0.06853867 0.04232376]
i = 20 -- x_k = [0.88635913 0.07003877 0.0436021 ]
i = 25 -- x_k = [0.88542672 0.07054218 0.0440311 ]
i = 30 -- x_k = [0.8851138  0.07071113 0.04417507]
i = 35 -- x_k = [0.88500879 0.07076782 0.04422338]
i = 40 -- x_k = [0.88497355 0.07078685 0.0442396 ]
i = 45 -- x_k = [0.88496173 0.07079324 0.04424504]
i = 50 -- x_k = [0.88495776 0.07079538 0.04424687]
i = 55 -- x_k = [0.88495642 0.0707961  0.04424748]
i = 60 -- x_k = [0.88495598 0.07079634 0.04424768]
i = 65 -- x_k = [0.88495583 0.07079642 0.04424775]
i = 70 -- x_k = [0.88495578 0.07079645 0.04424778]
i = 75 -- x_k = [0.88495576 0.07079646 0.04424778]
i = 80 -- x_k = [0.88495576 0.07079646 0.04424779]
i = 85 -- x_k = [0.88495575 0.07079646 0.04424779]
i = 90 -- x_k = [0.88495575 0.07079646 0.04424779]
i = 95 -- x_k = [0.88495575 0.07079646 0.04424779]


In [3]:
# Insert code here
x_k = np.matmul(np.linalg.matrix_power(A, i),x_0)
x_k = limit(x_k,i,oo)
print(x_k)

NameError: name 'i' is not defined

In [266]:
eig_val, eig_vec = np.linalg.eig(A)
D = np.diag(eig_val)
print(f'D =\n{D}\n')
P = eig_vec
print(f'P =\n{P}\n')
print(f'P Inverse =\n{P.I}\n')

A_k = P @ D**np.inf @ P.I
print(f'\nA_k =\n{A_k}')

D =
[[1.         0.         0.        ]
 [0.         0.80382635 0.        ]
 [0.         0.         0.25117365]]

P =
[[ 0.99557949  0.81563112  0.32258572]
 [ 0.07964636 -0.44036406 -0.8108726 ]
 [ 0.04977897 -0.37526706  0.48828688]]

P Inverse =
[[ 0.88888508  0.88888508  0.88888508]
 [ 0.13565548 -0.80459134 -1.42576345]
 [ 0.01363798 -0.70897751  0.86160445]]


A_k =
[[inf inf inf]
 [inf inf inf]
 [inf inf inf]]


#### Q3.

Does the long-term value of the probability distribution vector depend on the particular starting value $x^{(0)}$?

#### Answer:

No, the long-term value of the probability distribution vector does not depend on $x^{(0)}$ based on the codes below.

In [125]:
# Insert code here
x_0_test = np.array([[1/3, 1/3, 1/3],
                    [0.5,0.5,0],
                    [0.2,0.5,0.3]])

for j in x_0_test:
    x_k = np.matmul(np.linalg.matrix_power(A, i),j)
    x_k = limit(x_k,i,oo)
    print(f'x_0 = {j} -- x_k = {x_k}')

x_0 = [0.33333333 0.33333333 0.33333333] -- x_k = [0.884955751979813, 0.0707964603025572, 0.0442477877176238]
x_0 = [0.5 0.5 0. ] -- x_k = [0.884955752100978, 0.0707964602371396, 0.0442477876618766]
x_0 = [0.2 0.5 0.3] -- x_k = [0.884955751944952, 0.0707964603213788, 0.0442477877336631]


In [134]:
x_0_test = np.array([[1/3, 1/3, 1/3],
                     [0.5,0.5,0],
                     [0.2,0.5,0.3]])
for j in x_0_test:
    print(f'x_0 = {j}')
    for i in range(1,100):
        x_k = np.matmul(np.linalg.matrix_power(A, i),j)
        if i%5 == 0:
            print(f'i = {i} -- x_k = {x_k.T}')
    print('\n')

x_0 = [0.33333333 0.33333333 0.33333333]
i = 5 -- x_k = [0.69385395 0.17393818 0.13220788]
i = 10 -- x_k = [0.8208176  0.10542499 0.07375741]
i = 15 -- x_k = [0.86343151 0.08241752 0.05415096]
i = 20 -- x_k = [0.8777324  0.07469639 0.04757121]
i = 25 -- x_k = [0.88253165 0.07210524 0.0453631 ]
i = 30 -- x_k = [0.88414225 0.07123568 0.04462208]
i = 35 -- x_k = [0.88468275 0.07094386 0.0443734 ]
i = 40 -- x_k = [0.88486413 0.07084593 0.04428994]
i = 45 -- x_k = [0.88492501 0.07081306 0.04426193]
i = 50 -- x_k = [0.88494543 0.07080203 0.04425253]
i = 55 -- x_k = [0.88495229 0.07079833 0.04424938]
i = 60 -- x_k = [0.88495459 0.07079709 0.04424832]
i = 65 -- x_k = [0.88495536 0.07079667 0.04424797]
i = 70 -- x_k = [0.88495562 0.07079653 0.04424785]
i = 75 -- x_k = [0.88495571 0.07079648 0.04424781]
i = 80 -- x_k = [0.88495574 0.07079647 0.04424779]
i = 85 -- x_k = [0.88495575 0.07079646 0.04424779]
i = 90 -- x_k = [0.88495575 0.07079646 0.04424779]
i = 95 -- x_k = [0.88495575 0.07079646 0.0

#### Q4.

What is the value of $\lim_{k \rightarrow \infty} A^k$?

In [187]:
# Insert code here
A_k = np.linalg.matrix_power(A, 200)
print(A_k)

[[0.88495575 0.88495575 0.88495575]
 [0.07079646 0.07079646 0.07079646]
 [0.04424779 0.04424779 0.04424779]]


#### Q5.

Compute the eigenvalues and eigenvectors of $A$. 

In [160]:
# Insert code here
print(f'eigenvalues of A:\n {np.linalg.eig(A)[0]}\n\neigenvectors of A: \n{np.linalg.eig(A)[1]}')

eigenvalues of A:
 [1.         0.80382635 0.25117365]

eigenvectors of A: 
[[ 0.99557949  0.81563112  0.32258572]
 [ 0.07964636 -0.44036406 -0.8108726 ]
 [ 0.04977897 -0.37526706  0.48828688]]


#### Q6.

Explain your answer in <b>Q4</b> in terms of the eigenvalues and eigenvectors.

#### Answer:

Using Eigendecomposition, we can decompose $\lim_{k \rightarrow \infty} A^k$ into $\lim_{k \rightarrow \infty} A^k = \lim_{k \rightarrow \infty} P D^{k} P^{-1}$

In [217]:
# Insert code here
eig_val, eig_vec = np.linalg.eig(A)
D = np.matrix(np.diag(eig_val))
print(f'D =\n{D}\n')
P = np.matrix(eig_vec)
print(f'P =\n{P}\n')
print(f'P Inverse =\n{P.I}\n')

A_k = P * np.linalg.matrix_power(D, 200) * P.I
print(f'\nA_k =\n{A_k}')

D =
[[1.         0.         0.        ]
 [0.         0.80382635 0.        ]
 [0.         0.         0.25117365]]

P =
[[ 0.99557949  0.81563112  0.32258572]
 [ 0.07964636 -0.44036406 -0.8108726 ]
 [ 0.04977897 -0.37526706  0.48828688]]

P Inverse =
[[ 0.88888508  0.88888508  0.88888508]
 [ 0.13565548 -0.80459134 -1.42576345]
 [ 0.01363798 -0.70897751  0.86160445]]


A_k =
[[0.88495575 0.88495575 0.88495575]
 [0.07079646 0.07079646 0.07079646]
 [0.04424779 0.04424779 0.04424779]]


#### Q7.

Must 1 always be an eigenvalue of the transition matrix of a Markov chain? Why?

#### Answer:

$\lambda$ is an eigenvalue of $A$ if and only if: \
&emsp;&emsp;$\text{det}(A-\lambda • I) = 0$

The determinant of $A - 1 • I = 0$

Therefore, $\lambda = 1$ is always an eigenvalue of the transition matrix of a Markov chain.

#### Q8.

A probability distribution vector $x$ is said to be <b><i>stationary</i></b> if $Ax = x$. 

How can you determine such a stationary value $x$ using the eigenvalues and eigenvectors of $A$?

<i>Some Hints</i>:

1) Recall the formal definition of an eigenvalue-eigenvector pair. Can you see its relation to the question being asked?

2) Try solving <b>Q5</b> by hand <u>OR</u> try to understand what `numpy.linalg.eig` does to your eigenvectors.

#### Answer:

$x$ = eigenvector of $A$ where eigenvalue is 1
\begin{align}
\lambda = 1,\  
x = \begin{bmatrix}
0.99557949 \\
0.07964636 \\
0.04977897
\end{bmatrix}
\end{align}

In [244]:
# Insert code here
eig_val, eig_vec = np.linalg.eig(A)
x_stationary = eig_vec.T[0]
x_stationary = x_stationary.T

#Showing that Ax = x
print(f'x =\n{x_stationary}')
print('\n')
prod = np.matmul(A, x_stationary)

print(f'Ax =\n{prod}')

x =
[[0.99557949]
 [0.07964636]
 [0.04977897]]


Ax =
[[0.99557949]
 [0.07964636]
 [0.04977897]]


#### Q9.

How can you determine a stationary value $x$ <i>without</i> knowledge of the eigenvalues and eigenvectors of $A$?

#### Answer:

We can use

#### Q10.

Can there be more than one stationary distribution vector for a given Markov chain? If not, why? If so, give an example.

#### Answer:

(?)

In [ ]:
# Insert code here

#### Bonus.

Is it possible for a Markov chain to <b><u>NOT</b></u> have a stationary distribution? If not, why? If so, give an example.

#### Answer:

(?)